### ISSUE TITLE
Allow vim.notify to take a table of notifications

### ISSUE THREAD

**Issue by anott03** [neovim/neovim#15103]
All that would change is that usage could look something like this:
```lua
vim.notify({
    { "some message", log_level, opts }
    { "another message", a_different_log_level, other_opts }
}
```

This stems from a discussion we are having here: https://github.com/nvim-lua/plenary.nvim/pull/188. The general idea is that we split notifications into two components. For lack of a better term call them a "provider" and a "consumer." What would happen is that notifications are sent through `vim.notify`, the same way it is used right now, but then users could override that with a consumer to render notifications as they'd like. For more details see the discussion linked above.

@clason correct me if I've described anything inaccurately :)
CC @teto

**Comment by teto** [comment]
You should already be able to plug your own implementation with
```
lua vim.notify = function (...) print("toto") end
```
as for your example, you can do a loop of vim.notify ?

**Comment by clason** [comment]
Yes, but the idea was to extend the _API_ common to all such implementations to handle multiple concurrent notifications. Otherwise the producer (calling `vim.notify`) has to do a loop that the consumer (implementing `vim.notify`) has to unroll again. (The motivation is for something like https://neovim.discourse.group/t/wip-animated-notifications-plugin/448.)

**Comment by justinmk** [comment]
I do not understand why we added a half-baked nvim_notify + vim.notify + nvim_echo

**Comment by teto** [comment]
nvim_notify may not have been the best thought out features . For context, I wanted to be able to forward some messages as system notifications rather than via vim's `:messages` system due to its interruptive nature. A proper fix would be a rework of messages but that seemed unlikely. Until this happens we have with vim.notify a behavior easily overridable  that may be later aliased to `nvim_echo( msg,  { notify: true}`.

**Comment by justinmk** [comment]
Yeah. We could in those cases add an `opt` param to keep the door open for future improvements.
